File: server/src/skill/registry/skill_registry.py
from __future__ import annotations

from dataclasses import dataclass, field

from skill.base import SkillBase, SkillRequest, SkillResult


@dataclass(slots=True)
class SkillRegistry:
    _skills: dict[str, SkillBase] = field(default_factory=dict)

    def register(self, skill: SkillBase) -> None:
        if skill.name in self._skills:
            raise ValueError(f"Skill already registered: {skill.name}")
        self._skills[skill.name] = skill

    def get(self, name: str) -> SkillBase:
        skill = self._skills.get(name)
        if not skill:
            raise KeyError(f"Unknown skill: {name}")
        return skill

    def list_skills(self) -> list[dict[str, object]]:
        return [
            {
                "name": s.name,
                "description": s.description,
                "input_schema": s.input_schema,
                "output_schema": s.output_schema,
                "mode": s.mode.value,
            }
            for s in self._skills.values()
        ]

    def execute(self, name: str, request: SkillRequest) -> SkillResult:
        return self.get(name).execute(request)

File: phone/src/skill/registry/skill_registry.dart
import '../base/skill_base.dart';

class SkillRegistry {
  final Map<String, SkillBase> _skills = <String, SkillBase>{};

  void register(SkillBase skill) {
    if (_skills.containsKey(skill.name)) {
      throw StateError('Skill already registered: ${skill.name}');
    }
    _skills[skill.name] = skill;
  }

  SkillBase get(String name) {
    final skill = _skills[name];
    if (skill == null) {
      throw StateError('Unknown skill: $name');
    }
    return skill;
  }

  List<Map<String, dynamic>> listSkills() {
    return _skills.values
        .map(
          (skill) => <String, dynamic>{
            'name': skill.name,
            'description': skill.description,
            'input_schema': skill.inputSchema,
            'output_schema': skill.outputSchema,
            'mode': skill.mode,
          },
        )
        .toList(growable: false);
  }

  Future<Map<String, dynamic>> execute(String name, Map<String, dynamic> input) {
    return get(name).execute(input);
  }
}

